spacy/training/alignment_array.pyx

```diff
@@ -0,0 +1,66 @@
+from typing import List
+import numpy
+
+
+cdef class AlignmentArray:
+    """AlignmentArray is a simplified version of Ragged with two simplfications:
+    indexing returns numpy arrays and this type can only be used for CPU arrays.
+    However, these changes make AlginmentArray more efficient for indexing in a
+    tight loop."""
+
+    __slots__ = []
+
+    def __init__(self, alignment: List[List[int]]):
+        self._lengths = None
+        self._starts_ends = numpy.zeros(len(alignment) + 1, dtype="i")
+
+        cdef int data_len = 0
+        cdef int outer_len
+        cdef int idx
+        for idx, outer in enumerate(alignment):
+            outer_len = len(outer)
+            self._starts_ends[idx + 1] = self._starts_ends[idx] + outer_len
+            data_len += outer_len
+
+        self._data = numpy.empty(data_len, dtype="i")
+        idx = 0
+        for outer in alignment:
+            for inner in outer:
+                self._data[idx] = inner
+                idx += 1
+
+    def __getitem__(self, idx):
+        starts = self._starts_ends[:-1]
+        ends = self._starts_ends[1:]
+        if isinstance(idx, int):
+            start = starts[idx]
+            end = ends[idx]
+        elif isinstance(idx, slice):
```

Comment:
I just noticed that I don't throw an exception when the index is not an int or a slice. I'll add a commit to fix this.